File: examples/config-audit.ts
/*
 * TSN Config Audit
 *
 * Reads deployment-style key/value config from stdin and prints a readiness
 * summary.
 */

import { createRequire } from "module";
const require = createRequire(import.meta.url);

function readStdin(): string {
  const fs = require("fs");
  return fs.readFileSync("/dev/stdin", "utf-8");
}

function csvValues(raw: string): string[] {
  const trimmed: string = raw.trim()
  if (trimmed.length === 0) return []
  const values: string[] = []
  for (const part of trimmed.split(",")) {
    const value: string = part.trim().toLowerCase()
    if (value.length > 0) values.push(value)
  }
  return values
}

function boolText(value: boolean): string {
  return value ? "true" : "false"
}

function main(): void {
  const config = new Map<string, string>()
  for (const rawLine of readStdin().split("\n")) {
    const line: string = rawLine.trim()
    if (line.length === 0 || line.startsWith("#")) continue
    const eq: number = line.indexOf("=")
    if (eq === -1) continue
    config.set(line.slice(0, eq).trim().toUpperCase(), line.slice(eq + 1).trim())
  }

  const env: string = (config.get("APP_ENV") ?? "").toUpperCase()
  const featureFlags: string[] = csvValues(config.get("FEATURE_FLAGS") ?? "")
  const origins: string[] = csvValues(config.get("ALLOWED_ORIGINS") ?? "")
  const required: string[] = [
    "APP_ENV",
    "API_BASE_URL",
    "ENABLE_AUDIT_LOG",
    "SESSION_COOKIE_SECURE",
  ]

  const hasAllRequired: boolean = required.every((key: string): boolean => (config.get(key) ?? "").length > 0)
  const firstMissingIdx: number = required.findIndex((key: string): boolean => (config.get(key) ?? "").length === 0)
  const auditEnabled: boolean = (config.get("ENABLE_AUDIT_LOG") ?? "").toLowerCase() === "true"
  const secureCookies: boolean = (config.get("SESSION_COOKIE_SECURE") ?? "").toLowerCase() === "true"
  const hasRiskyFlags: boolean = featureFlags.some((flag: string): boolean => {
    return flag.startsWith("alpha") || flag.endsWith("_dev") || flag.includes("debug")
  })
  const originsSafe: boolean = origins.every((origin: string): boolean => {
    return !origin.includes("localhost") && !origin.includes("127.0.0.1") && !origin.startsWith("http://")
  })

  console.log("=== CONFIG AUDIT ===")
  console.log("Environment: " + env)
  console.log("Required complete: " + boolText(hasAllRequired))
  console.log("First missing required key: " + (firstMissingIdx === -1 ? "none" : required[firstMissingIdx]))
  console.log("Audit logging enabled: " + boolText(auditEnabled))
  console.log("Secure cookies enabled: " + boolText(secureCookies))
  console.log("Risky flags present: " + boolText(hasRiskyFlags))
  console.log("Origins production-safe: " + boolText(originsSafe))
  console.log("Feature flags: " + featureFlags.join(", "))
}

main()

File: harness/rust/config-audit.rs
// Rust port of examples/config-audit.ts — zero-dep, stdlib only.

use std::io::{self, Read, Write, BufWriter};

struct Entry {
    key: String,
    value: String,
}

fn parse_entries(raw: &str) -> Vec<Entry> {
    let mut entries = Vec::new();
    for line in raw.split('\n') {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let eq = match line.find('=') {
            Some(i) => i,
            None => continue,
        };
        let key = line[..eq].trim().to_uppercase();
        let value = line[eq + 1..].trim().to_string();
        entries.push(Entry { key, value });
    }
    entries
}

fn value_for<'a>(entries: &'a [Entry], key: &str) -> &'a str {
    for e in entries {
        if e.key == key {
            return &e.value;
        }
    }
    ""
}

fn first_missing_idx(entries: &[Entry], required: &[&str]) -> i32 {
    for (i, key) in required.iter().enumerate() {
        if value_for(entries, key).is_empty() {
            return i as i32;
        }
    }
    -1
}

fn csv_values(raw: &str) -> Vec<String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Vec::new();
    }
    let mut out = Vec::new();
    for part in trimmed.split(',') {
        let v = part.trim().to_lowercase();
        if !v.is_empty() {
            out.push(v);
        }
    }
    out
}

fn bool_text(v: bool) -> &'static str { if v { "true" } else { "false" } }

fn main() {
    let mut raw = String::new();
    io::stdin().read_to_string(&mut raw).unwrap();

    let entries = parse_entries(&raw);
    let env = value_for(&entries, "APP_ENV").to_uppercase();
    let feature_flags = csv_values(value_for(&entries, "FEATURE_FLAGS"));
    let origins = csv_values(value_for(&entries, "ALLOWED_ORIGINS"));
    let required = ["APP_ENV", "API_BASE_URL", "ENABLE_AUDIT_LOG", "SESSION_COOKIE_SECURE"];

    let has_all = required.iter().all(|k| !value_for(&entries, k).is_empty());
    let first_miss = first_missing_idx(&entries, &required);
    let audit = value_for(&entries, "ENABLE_AUDIT_LOG").to_lowercase() == "true";
    let secure = value_for(&entries, "SESSION_COOKIE_SECURE").to_lowercase() == "true";
    let risky = feature_flags.iter().any(|f| f.starts_with("alpha") || f.ends_with("_dev") || f.contains("debug"));
    let safe = origins.iter().all(|o| !o.contains("localhost") && !o.contains("127.0.0.1") && !o.starts_with("http://"));

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    writeln!(out, "=== CONFIG AUDIT ===").unwrap();
    writeln!(out, "Environment: {}", env).unwrap();
    writeln!(out, "Required complete: {}", bool_text(has_all)).unwrap();
    let missing_label = if first_miss == -1 { "none".to_string() } else { required[first_miss as usize].to_string() };
    writeln!(out, "First missing required key: {}", missing_label).unwrap();
    writeln!(out, "Audit logging enabled: {}", bool_text(audit)).unwrap();
    writeln!(out, "Secure cookies enabled: {}", bool_text(secure)).unwrap();
    writeln!(out, "Risky flags present: {}", bool_text(risky)).unwrap();
    writeln!(out, "Origins production-safe: {}", bool_text(safe)).unwrap();
    writeln!(out, "Feature flags: {}", feature_flags.join(", ")).unwrap();
}
